### PR TITLE
Added the option to pass arguments to add-on menu callbacks

### PR DESCRIPTION
**Pull request by basrieter**
## Description

I added the option to pass arguments to add-on menu callbacks.

## Motivation and Context

The addon.xml allows the configuration of context menu items. However, only a `library` parameter can be passed to such menu items:

```xml
<item library="menu_settings.py">
    <label>30528</label>
</item>
```

With the current implementation it is only possible to specify a `library`, so each menu item needs a separate Python library for callback. This requires a single Python file for each menu item. This PR introduces an `args` property:

```xml
<item library="menu.py" args="?action=settings">
    <label>30528</label>
</item>
```

That property allows you to pass an argument to the script:

```
sys.argv=['plugin.video.pluginid', '?action=settings']
```

These arguments are similar to those passed to normal plugin runs, the only difference is that there is no _handle_ passed. 

This PR also enables the `reuselanguageinvoker` for menu callbacks. This was not there.

## How Has This Been Tested?

Locally on latest sources with my add-on (Retrospect) containing context menu items.

## Screenshots (if appropriate):
n.a.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/ContextMenuItem.cpp [hunk 1]
```diff
@@ -50,9 +50,13 @@ bool CContextMenuItem::Execute(const CFileItemPtr& item) const
   if (!CServiceBroker::GetAddonMgr().GetAddon(m_addonId, addon))
     return false;
 
+  bool reuseLanguageInvoker = false;
+  if (addon->ExtraInfo().find("reuselanguageinvoker") != addon->ExtraInfo().end())
+    reuseLanguageInvoker = addon->ExtraInfo().at("reuselanguageinvoker") == "true";
+
```
Comment (Montellese) [inline xbmc/ContextMenuItem.cpp:56]: Move the `reuselanguageinvoker` stuff into a separate commit.

**Review by Montellese — CHANGES_REQUESTED** [review]
Great idea. I was just playing around with context menu items recently and found it very annoying that I had to create a separate python file for every item.

I made a few suggestions.

**Comment by Montellese** [comment]
Do we check somewhere in the code that the `addon.xml` doesn't contain the same script for multiple context menu items? Or how is this restriction handled?

**Comment by enen92** [comment]
@Montellese we don't but we can definetly add that to https://github.com/xbmc/addon-check addon.xml schema validators (for matrix and above) to enforce the use of args

File: xbmc/ContextMenuItem.h [hunk 2]
```diff
@@ -69,8 +69,9 @@ class CContextMenuItem : public IContextMenuItem
     const std::string& parent,
     const std::string& library,
     const std::string& condition,
-    const std::string& addonId);
-
+    const std::string& addonId, 
+    const std::vector<std::string>& args);
```
[the remark this answers is not in the document]
Comment (basrieter) [reply xbmc/ContextMenuItem.h]: Done

File: xbmc/addons/ContextMenuAddon.cpp [hunk 3]
```diff
@@ -87,10 +87,19 @@ void CContextMenuAddon::ParseMenu(
     if (StringUtils::IsNaturalNumber(label))
       label = g_localizeStrings.GetAddonString(ID(), atoi(label.c_str()));
 
+    std::vector<std::string> args;
+    args.push_back(ID());
+ 
+    std::string arg = element.second.GetValue("@args").asString();
+    if (!args.empty())
+    {
```
[the remark this answers is not in the document]
Comment (basrieter) [reply xbmc/addons/ContextMenuAddon.cpp]: I always teach my students to include them, no matter what :) Done

File: xbmc/ContextMenuItem.cpp (hunk 1 above)
Comment (basrieter) [reply xbmc/ContextMenuItem.cpp:56]: Done.

**Comment by basrieter** [comment]
> Do we check somewhere in the code that the `addon.xml` doesn't contain the same script for multiple context menu items? Or how is this restriction handled?

We do check that somewhere, let me look into that. I should take the new `args` into account.

File: xbmc/addons/ContextMenuAddon.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Montellese) [reply xbmc/addons/ContextMenuAddon.cpp]: Personally I favour the following style
```cpp
if (arg.empty() == false) {
  args.push_back(arg);
}
```
but I thought our coding style requires no curly braces on single line blocks. Maybe I'm wrong?

File: xbmc/addons/ContextMenuAddon.cpp [hunk 4]
```diff
@@ -87,10 +87,17 @@ void CContextMenuAddon::ParseMenu(
     if (StringUtils::IsNaturalNumber(label))
       label = g_localizeStrings.GetAddonString(ID(), atoi(label.c_str()));
 
+    std::vector<std::string> args;
+    args.push_back(ID());
+ 
+    std::string arg = element.second.GetValue("@args").asString();
+    if (!args.empty())
```
Comment (Montellese) [inline xbmc/addons/ContextMenuAddon.cpp:94]:
```suggestion
    if (!arg.empty())
```

Comment (basrieter) [reply xbmc/addons/ContextMenuAddon.cpp]: Just found that one. I also update the `==` operator.

File: xbmc/addons/ContextMenuAddon.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (basrieter) [reply xbmc/addons/ContextMenuAddon.cpp]: Nope it does not. So now worries.

**Comment by basrieter** [comment]
> @Montellese we don't but we can definetly add that to https://github.com/xbmc/addon-check addon.xml schema validators (for matrix and above) to enforce the use of args

@enen92 Kodi already filters out duplicate entries. It was based on the `library` only, and I extended that to also take the `args` into consideration now. If the combination is a duplicate, only one will be shown. 

But it would be a good idea to check for this. If we merge this, the addon checker also needs an update to the schema of the addon.xml

**Comment by basrieter** [comment]
BTW: why did Travis CI not re-check this PR?

**Review by Montellese — APPROVED** [review]
Looks good to me now.

**Comment by basrieter** [comment]
jenkins build this please

**Comment by basrieter** [comment]
jenkins build this please

**Comment by ronie** [comment]
thx!

**Pull request merged**